kvserver: add option to skip merge queue if external files are present

stevendanna opened the pull request:
Replicas with external files are expensive to iterate. Further, they have very in-accurate stats. Here we add an option to skip the merge queue if a replica happens to have external files.

Fixes #120256

Release note: None

stevendanna committed:
kvserver: add option to skip merge queue if external files are present

Replicas with external files are expensive to iterate. Further, they
have very inaccurate stats. Here we add an option to skip the merge
queue if a replica happens to have external files.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120541)
<!-- Reviewable:end -->

dt commented on pkg/kv/kvserver/merge_queue.go:
Infof?

stevendanna commented:
@nvanbenschoten @miraradeva I believe in our recent discussion y'all indicated that something like this were we completely exclude some replica's from a queue would be a bit more tenable. Here we exclude replicas with external bytes from the merge queue.

miraradeva reviewed (commented):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLdceMA10Nwrz-471J):*
nit: Replica's -> Replicas 
___
*[`-- commits` line 5 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLdhwbE1shzNibG8sS):*
nit: inaccurate
___
*[`pkg/kv/kvserver/merge_queue.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLe9UZ0kl471FZYtc2:-NtLe9UZ0kl471FZYtc3:b9dvq3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue.go#L57)):*
> ```Go
> 
> // skipMergeQueueForExternalBytes is a setting that controls whether replicas
> // with external bytes are should be processed by the merge queue.
> ```

nit: remove "are"
___
*[`pkg/kv/kvserver/replica.go` line 1044 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLeQzD6uyhi0qOALzu:-NtLeQzD6uyhi0qOALzv:b-o57kck) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1044)):*
> ```Go
> }
> 
> // HasExternalBytes returns the number of external bytes in ther
> ```

It seems to return a boolean, not a number of bytes.
___
*[`pkg/kv/kvserver/replica.go` line 1050 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLfX6e66jjk32iK2VZ:-NtLfX6e66jjk32iK2V_:bh1j1h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1050)):*
> ```Go
> 	desc := r.Desc()
> 	sp := desc.KeySpan().AsRawSpanWithNoLocals()
> 	_, _, externalBytes, err := r.store.StateEngine().ApproximateDiskBytes(sp.Key, sp.EndKey)
> ```

Is `ApproximateDiskBytes` expensive? Making sure it's ok to call on all replicas considered by the merge queue?
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLkVA65BaNLUsora54:-NtLkVA75snVRYCig56m:b-izjzes) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L181)):*
> ```Go
> 
> // TestMergeWithExternalFiles tests that while a range has external
> // file bytes, we calculate an estimate of the stats during splits.
> ```

This is not testing estimated stats, right?
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLgdkM5VzqWwQM6kxd:-NtLgdkM5VzqWwQM6kxe:bz41nw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L182)):*
> ```Go
> // TestMergeWithExternalFiles tests that while a range has external
> // file bytes, we calculate an estimate of the stats during splits.
> func TestMergeShouldQueueWithExternalFiles(t *testing.T) {
> ```

`merge_queue_test.go` seems to contain only unit tests for enqueuing on the merge queue. I think you can do something similar because the only merge queue aspect your change touches is `shouldQueue`.  Is there a way to mock the external files or the response of `ApproximateDiskBytes(...)`, rather than actually starting a server and running `AddSST`?

Alternatively, if you wanted to add an integration test, `client_merge_test.go` is the place to do that. There you can actually let the merge queue go through with the merge and assert on the merged range.
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLiyeEClCo5lmFZZR6:-NtLiyeEClCo5lmFZZR7:b-v4yhxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L200)):*
> ```Go
> 			})
> 
> 			_, err := db.Exec("SET CLUSTER SETTING kv.range_merge.skip_external_bytes=$1", skipExternal)
> ```

This can be set as a cluster setting before starting the server above, right? Seems less prone to racing with the rest of the test.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-:-NtMCosqD606fqiA1Udv:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @miraradeva)

___
*[`pkg/kv/kvserver/merge_queue.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120541#gh-1526254037:-NtLrKClE9DkNy6li1Rq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f75b324cce85974c5b2149e5a41402011ac8cbbb/pkg/kv/kvserver/merge_queue.go#L62)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

True?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/merge_queue.go` line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120541#gh-1526254444:-NtLrLVg9wCNJMSQ5x8M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f75b324cce85974c5b2149e5a41402011ac8cbbb/pkg/kv/kvserver/merge_queue.go#L182)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Infof?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica.go` line 1044 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLeQzD6uyhi0qOALzu:-NtLrb3u3fQZdugP8dC6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1044)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

It seems to return a boolean, not a number of bytes.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica.go` line 1050 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLfX6e66jjk32iK2VZ:-NtLrfl02iJTlUH_NT6d:b-rey6m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1050)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Is `ApproximateDiskBytes` expensive? Making sure it's ok to call on all replicas considered by the merge queue?
</blockquote></details>

It's not free. It does involve iterating through pebble's metadata about the files in its store. However, for this reason we only call it when we were otherwise going to return `true` in shouldQueue.
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLkVA65BaNLUsora54:-NtLschD7g5JIgitorbm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L181)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This is not testing estimated stats, right?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLgdkM5VzqWwQM6kxd:-NtMCOZT0kCl0XE5-WtF:bq2wr0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L182)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

`merge_queue_test.go` seems to contain only unit tests for enqueuing on the merge queue. I think you can do something similar because the only merge queue aspect your change touches is `shouldQueue`.  Is there a way to mock the external files or the response of `ApproximateDiskBytes(...)`, rather than actually starting a server and running `AddSST`?

Alternatively, if you wanted to add an integration test, `client_merge_test.go` is the place to do that. There you can actually let the merge queue go through with the merge and assert on the merged range.
</blockquote></details>

Cool, I think I like the integration test a bit better here so I've moved this to client_merge_test.go
___
*[`pkg/kv/kvserver/merge_queue_test.go` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLiyeEClCo5lmFZZR6:-NtLse9e0YLczWN5TfVm:b-7as5qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/merge_queue_test.go#L200)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This can be set as a cluster setting before starting the server above, right? Seems less prone to racing with the rest of the test.
</blockquote></details>

Doneburger.

As a note, there should be no race in single-node tests for cluster settings since SET CLUSTER SETTING waits for the local node to see the setting update, but this protects us against future changes so it still makes sense to me.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @miraradeva)

___
*[`pkg/kv/kvserver/replica.go` line 1050 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLfX6e66jjk32iK2VZ:-NtMDH3HD8SxUqw7Ucbo:b-d6yu7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1050)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

It's not free. It does involve iterating through pebble's metadata about the files in its store. However, for this reason we only call it when we were otherwise going to return `true` in shouldQueue.
</blockquote></details>

We could alternatively push this into the `process` function to make sure we really only call this on ranges that we were about to merge anyway. If we are about to do an expensive merge, I'm less worried about calling this function. Thoughts?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
Something else Andrew mentioned here is that we can utilize the [queue config](https://github.com/cockroachdb/cockroach/blob/4a3a031ec225f56180f6883c2b5787203d835647/pkg/kv/kvserver/queue.go#L299) to do this check at the generic queue level, as opposed to each individual queue that we want to skip processing ranges with external data. Then in each individual queue, we just indicate if this queue should skip such ranges.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-:-NtMWw539Q_Am_eCP9pC:bwjkavu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[`pkg/kv/kvserver/replica.go` line 1050 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtLfX6e66jjk32iK2VZ:-NtMWFSeEvMs73Lo_aU-:bevjevu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e067fbc0dd43afd707f48420f0bce3415e4d9e7c/pkg/kv/kvserver/replica.go#L1050)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We could alternatively push this into the `process` function to make sure we really only call this on ranges that we were about to merge anyway. If we are about to do an expensive merge, I'm less worried about calling this function. Thoughts?
</blockquote></details>

I think `shouldQueue` is the right place for this logic because it will allow us to circumvent this check (by setting `skipShouldQueue` in `Enqueue()`) if we ever need to. Maybe we can run some tests to check the impact of this check on the queue processing times. 


<!-- Sent from Reviewable.io -->

stevendanna committed:
kvserver: move external bytes check to base queue

Moving it to the base queue will make it a little easier to opt other
queues into this feature.

For the moment, I've left this conditional on a setting and have not
added settings for all queues.

Epic: none
Release note: None

stevendanna reviewed (commented):
I've added a second commit that moves this to the base queue. I've also disable this by default until we can do a bit more testing on the cost of calling this.

If you are OK with this off-by-default I think it'll be easier to put together some good tests with it committed since we have a number of pieces of work in flight at the moment.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-:-NtQQBzm-8jXLM1P3BFx:bu2bex5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @miraradeva)



<!-- Sent from Reviewable.io -->

miraradeva reviewed (approved):
> Reviewed 2 of 4 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[`pkg/kv/kvserver/queue.go` line 353 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtQk9rr4tKF1hWAa0pv:-NtQk9rr4tKF1hWAa0pw:b-ibcv61) ([raw file](https://github.com/cockroachdb/cockroach/blob/540cd1c2fd243d6310aea655aea9ccbe2546a370/pkg/kv/kvserver/queue.go#L353)):*
> ```Go
> 	// skipIfReplicaHasExternalFilesConfig is a reference to the
> 	// clsuter setting that controls whether replicas should be
> 	// processed in this queueu if they have external files. May
> ```

nit: queue
___
*[`pkg/kv/kvserver/queue.go` line 357 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtQkDw8Dgjjt-lRARSR:-NtQkDw8Dgjjt-lRARSS:bd3693i) ([raw file](https://github.com/cockroachdb/cockroach/blob/540cd1c2fd243d6310aea655aea9ccbe2546a370/pkg/kv/kvserver/queue.go#L357)):*
> ```Go
> 	//
> 	// skipIfReplicaHasExternalFilesConfig is only consulted after
> 	// shouldQueue rturns true. for the given replica.
> ```

nit: returns true for
___
*[`pkg/kv/kvserver/queue.go` line 696 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtQkmyu16l-zuxq0xpn:-NtQkmyu16l-zuxq0xpo:b-8zcegm) ([raw file](https://github.com/cockroachdb/cockroach/blob/540cd1c2fd243d6310aea655aea9ccbe2546a370/pkg/kv/kvserver/queue.go#L696)):*
> ```Go
> 			log.Warningf(ctx, "could not determine if %s has external bytes: %s", realRepl, err)
> 			return
> 		} else if hasExternal {
> ```

nit: does this need to be `else if`? Can it be just a separate `if`?


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Thanks for the review!  I'll put together some numbers on how much this costs once a few other PRs land. Might not be until next week, but I don't think we need this enabled by default for 24.1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-:-NtQqM3NCqxHEUgIdJBu:b-8kaxwz)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @miraradeva)

___
*[`pkg/kv/kvserver/queue.go` line 696 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120541#-NtQkmyu16l-zuxq0xpn:-NtQqBUl0uJ7yVvjhTm6:b-ahxn4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/540cd1c2fd243d6310aea655aea9ccbe2546a370/pkg/kv/kvserver/queue.go#L696)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: does this need to be `else if`? Can it be just a separate `if`?
</blockquote></details>

Done. I also prefer this style but so many places in the code prefer leaving `err` out of scope I defaulted to that. But in this case it _really_ doesn't matter since we have an err in scope already.


<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=miraradeva
